_search-plugins/knn/filter-search-knn.md

```diff
@@ -0,0 +1,519 @@
+
+# Search for k-NN with filters
+Introduced 2.4
+{: .label .label-purple }
+
+You can create custom filters using Query DSL search options to refine your k-NN searches. You define the filter criteria within the `knn_vector` field's `filter` subsection in your query. You can use all available Query DSL query types such as: term, range, regex and wildcard. To include or exclude results, you specify Boolean query clauses. You also specify a query point with the `knn_vector` type and search for nearest neighbors that match your filter criteria.
+To run k-NN queries with a filter, the Lucene search engine and HSNW method are required.
+
+To learn more about how to use Query DSL Boolean query clauses, see [Boolean queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/bool).
+{: .note }
+
+## How does a k-NN filter work?
+
+Starting from OpenSearch version 2.2 the k-NN plugin uses the Lucene engine that provides a search process for filtered documents that uses a Hierarchical Navigable Small World (HSNW) algorithm to represent a multi-layered graph to filter searches. After a filter is applied to a set of documents to be searched, the algorithm decides whether to perform pre-filtering for an exact kNN search or modified post-filtering for approximate search. The approximate search with filtering guarantees the top number of closest vectors in result.
+
+Lucene also provides the capability to operate its `KnnVectorQuery` over a subset of documents. To learn more about Lucene’s new capability, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).
+
+### Filtered search performance
+
+The Lucene engine and HSNW algorithm allow you to to apply k-NN searches more efficiently both in terms of relevancy of search results and performance. Consider that if you do an exact search on a large data set, the results are slow, and post-filtering does not guarantee the required number of results you specify for the `k` value.
+With this new capability, you can create an approximate k-NN search and apply filters, with the amount of results you need.
+
+The following workflow diagram shows how the HSNW algorithm decides which type of filtering to apply to a search based on the volume of documents, and number of `k` points in the index that you search with a filter.
+
+![How the algorithm evaluates a doc set]({{site.url}}{{site.baseurl}}/images/hsnw-algorithm.png)
+
+***Figure 1: Filter algorithm workflow***
+
+## Filter approaches by use case
+
+Depending on the data set that you are searching, you might choose a different approach to minimize recall or latency. You can create filters that are either: very selective (80%), somewhat selective (38%), or not very selective (2.5%). The selectiveness percentage indicates the amount the filter returns for any given document set in an index.
+
+In this context, `score_script` is essentially a brute force search, whereas boolean filter is an approximate k-NN search with post filtering.
```

Comment:
I think we may need to add more context around `scoring script` and `boolean filter` approaches. They both are available for users today.For scoring script we can provide a link https://opensearch.org/docs/latest/search-plugins/knn/knn-score-script/. I can't find existing link for boolean query post filtering, maybe we can say that in such case knn query joined with a filter through `bool` `must` clause, or provide an example of such query:
```
{
    "size": 3,
    "query": {
        "bool": {
            "filter": {
                "bool": {
                    "must": [
                        {
                            "range": {
                                "rating": {
                                    "gte": 8,
                                    "lte": 10
                                }
                            }
                        },
                        {
                            "term": {
                                "parking": "true"
                            }
                        }
                    ]
                }
            },
            "must": [
                {
                    "knn": {
                        "location": {
                            "vector": [
                                5.0,
                                4.0
                            ],
                            "k": 20
                        }
                    }
                }
            ]
        }
    }
}
```